### PR TITLE
SQL DOM Parameterized Query 

### PR DESCRIPTION
**Pull request by bchong95**
# SQL DOM Parameterized Query 

## Description

Adds support for parameterized queries in the SQL DOM. Needed for LINQ parameterization 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/Core/Number64.cs [hunk 1]
```diff
@@ -11,6 +11,7 @@ namespace Microsoft.Azure.Cosmos
     /// <summary>
     /// Struct that represents either a double or 64 bit int
     /// </summary>
+    [JsonConverter(typeof(Number64JsonConverter))]
 #if INTERNAL
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/Number64.cs]: Please fix the failing tests

File: Microsoft.Azure.Cosmos/src/SqlObjects/SqlParameter.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+﻿// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ------------------------------------------------------------
+namespace Microsoft.Azure.Cosmos.Sql
+{
+    using System;
+
+    internal sealed class SqlParameter : SqlObject
```
Comment (simplynaveen20) [inline Microsoft.Azure.Cosmos/src/SqlObjects/SqlParameter.cs]: We already have SqlParameter in V3, which is basically name value pair to hold the parameters. Should we rename this one ?

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/SqlObjects/SqlParameter.cs]: They are different namespaces, so I am going to leave as is to reflect the naming of the grammar.

File: Microsoft.Azure.Cosmos/src/Linq/QueryUnderConstruction.cs [hunk 3]
```diff
@@ -628,6 +693,34 @@ public QueryUnderConstruction AddTopSpec(SqlTopSpec topSpec)
...
+                throw new ArgumentException($"Expected number literal scalar expression.");
+            }
+
+            SqlLiteral sqlLiteral = topLiteralExpression.Literal;
+            if (!(sqlLiteral is SqlNumberLiteral sqlNumberLiteral))
+            {
+                throw new ArgumentException($"Expected number literal.");
+            }
+
+            if (!sqlNumberLiteral.Value.IsInteger)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Linq/QueryUnderConstruction.cs]: Is there any possibility that value can be null?

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Linq/QueryUnderConstruction.cs]: The value is a struct.

File: Microsoft.Azure.Cosmos/src/Query/Core/Number64.cs (hunk 1 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Number64.cs]: There was a bug in how core was doing double.NaN.GetHashCode().

**Pull request merged**